x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/TimeseriesLifecycleActionsRegistry.java

```diff
@@ -0,0 +1,143 @@
+            ShrinkAction.NAME,
+            ForceMergeAction.NAME,
+            SearchableSnapshotAction.NAME
+        ).filter(Objects::nonNull).toList()
+    );
+
+    public static final Map<Integer, List<String>> ORDERED_VALID_WARM_ACTIONS = Map.of(
+        VERSION_ONE,
+        Stream.of(
+            SetPriorityAction.NAME,
+            UnfollowAction.NAME,
+            ReadOnlyAction.NAME,
+            DownsampleAction.NAME,
+            AllocateAction.NAME,
+            MigrateAction.NAME,
+            ShrinkAction.NAME,
+            ForceMergeAction.NAME
+        ).filter(Objects::nonNull).toList(),
+        VERSION_TWO,
+        Stream.of(
+            SetPriorityAction.NAME,
+            UnfollowAction.NAME,
+            ReadOnlyAction.NAME,
+            AllocateAction.NAME,
+            MigrateAction.NAME,
+            DownsampleAction.NAME,
+            ShrinkAction.NAME,
+            ForceMergeAction.NAME
+        ).filter(Objects::nonNull).toList()
+    );
+
+    public static final Map<Integer, List<String>> ORDERED_VALID_COLD_ACTIONS = Map.of(
+        VERSION_ONE,
+        Stream.of(
+            SetPriorityAction.NAME,
+            UnfollowAction.NAME,
+            ReadOnlyAction.NAME,
+            DownsampleAction.NAME,
+            SearchableSnapshotAction.NAME,
+            AllocateAction.NAME,
+            MigrateAction.NAME,
+            FreezeAction.NAME
+        ).filter(Objects::nonNull).toList(),
+        VERSION_TWO,
+        Stream.of(
+            SetPriorityAction.NAME,
+            UnfollowAction.NAME,
+            ReadOnlyAction.NAME,
+            SearchableSnapshotAction.NAME,
+            AllocateAction.NAME,
+            MigrateAction.NAME,
+            DownsampleAction.NAME,
+            FreezeAction.NAME
+        ).filter(Objects::nonNull).toList()
+    );
+
+    public static final Map<Integer, List<String>> ORDERED_VALID_FROZEN_ACTIONS = Map.of(
+        VERSION_ONE,
+        List.of(UnfollowAction.NAME, SearchableSnapshotAction.NAME),
+        VERSION_TWO,
+        List.of(UnfollowAction.NAME, SearchableSnapshotAction.NAME)
+    );
+    public static final Map<Integer, List<String>> ORDERED_VALID_DELETE_ACTIONS = Map.of(
+        VERSION_ONE,
+        List.of(WaitForSnapshotAction.NAME, DeleteAction.NAME),
+        VERSION_TWO,
+        List.of(WaitForSnapshotAction.NAME, DeleteAction.NAME)
+    );
+
+    static final Set<String> VALID_HOT_ACTIONS = Sets.newHashSet(ORDERED_VALID_HOT_ACTIONS.get(CURRENT_VERSION));
+    static final Set<String> VALID_WARM_ACTIONS = Sets.newHashSet(ORDERED_VALID_WARM_ACTIONS.get(CURRENT_VERSION));
+    static final Set<String> VALID_COLD_ACTIONS = Sets.newHashSet(ORDERED_VALID_COLD_ACTIONS.get(CURRENT_VERSION));
+    static final Set<String> VALID_FROZEN_ACTIONS = Sets.newHashSet(ORDERED_VALID_FROZEN_ACTIONS.get(CURRENT_VERSION));
+    static final Set<String> VALID_DELETE_ACTIONS = Sets.newHashSet(ORDERED_VALID_DELETE_ACTIONS.get(CURRENT_VERSION));
```

Comment:
Are we concerned that a step may be considered valid but that a different version may not contain the action name? For example, we have to be careful not to remove any actions from later version because they will be considered "invalid" even if they exist in older action version orders.

Should we have an assert for this?